csharp/src/Drivers/Apache/ApacheUtility.cs

```diff
@@ -0,0 +1,141 @@
+
+using System;
+using System.Threading;
+
+namespace Apache.Arrow.Adbc.Drivers.Apache
+{
+    internal class ApacheUtility
+    {
+        internal const int QueryTimeoutSecondsDefault = 60;
+
+        public enum TimeUnit
+        {
+            Seconds,
+            Milliseconds
+        }
+
+        public static CancellationToken GetCancellationToken(int timeout, TimeUnit timeUnit)
+        {
+            TimeSpan span;
+
+            if (timeout == 0 || timeout == int.MaxValue)
+            {
+                // the max TimeSpan for CancellationTokenSource is int.MaxValue in milliseconds (not TimeSpan.MaxValue)
+                // no matter what the unit is
+                span = TimeSpan.FromMilliseconds(int.MaxValue);
+            }
+            else
+            {
+                if (timeUnit == TimeUnit.Seconds)
+                {
+                    span = TimeSpan.FromSeconds(timeout);
+                }
+                else
+                {
+                    span = TimeSpan.FromMilliseconds(timeout);
+                }
+            }
+
+            return GetCancellationToken(span);
+        }
+
+        private static CancellationToken GetCancellationToken(TimeSpan timeSpan)
+        {
+            var cts = new CancellationTokenSource(timeSpan);
+            return cts.Token;
+        }
+
+        public static bool QueryTimeoutIsValid(string key, string value, out int queryTimeoutSeconds)
+        {
+            if (!string.IsNullOrEmpty(value) && int.TryParse(value, out int queryTimeout) && (queryTimeout >= 0))
+            {
+                queryTimeoutSeconds = queryTimeout;
+                return true;
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException(key, value, $"The value '{value}' for option '{key}' is invalid. Must be a numeric value of 0 (infinite) or greater.");
+            }
+        }
+
+        public static bool ContainsException<T>(Exception? exception, out T? containedException) where T : Exception
+        {
+            if (exception is AggregateException aggregateException)
+            {
+                foreach (Exception? ex in aggregateException.InnerExceptions)
+                {
+                    if (ex is T ce)
+                    {
+                        containedException = ce;
+                        return true;
+                    }
+                }
+            }
+
+            Exception? e = exception;
+            while (e != null)
+            {
+                if (e is T ce)
+                {
+                    containedException = ce;
+                    return true;
+                }
+                e = e.InnerException;
+            }
+
+            containedException = null;
+            return false;
+        }
+
+        public static bool ContainsException(Exception? exception, Type? exceptionType, out Exception? containedException)
+        {
+            if ( exception == null || exceptionType == null)
```

Comment:
nit: extra space after parenthesis

If you're using Visual Studio, you can trigger the VS autoformat by doing Control-K Control-D.